### PR TITLE
Respect doc(hidden) when suggesting available fields

### LINKED ISSUE #93210: 'available fields are' message in diagnostic lists unstable/hidden fields.
With this patch to std:

```diff
diff --git a/library/core/src/pin.rs b/library/core/src/pin.rs
index ce5ed40940ec6..b8511eb664cb5 100644
--- a/library/core/src/pin.rs
+++ b/library/core/src/pin.rs
@@ -406,7 +406,9 @@ use crate::ops::{CoerceUnsized, Deref, DerefMut, DispatchFromDyn, Receiver};
 #[repr(transparent)]
 #[derive(Copy, Clone)]
 pub struct Pin<P> {
-    pointer: P,
+    #[unstable(feature = "pin_internals", issue = "none")]
+    #[doc(hidden)]
+    pub pointer: P,
 }
```

The hidden `.pointer` field appears as a suggestion in a diagnostic:

```
---- [ui] ui/async-await/suggest-switching-edition-on-await.rs stdout ----
diff of stderr:

24	LL |     x.await;
25	   |       ^^^^^ unknown field
26	   |
+	   = note: available fields are: `pointer`
27	   = note: to `.await` a `Future`, switch to Rust 2018 or later
28	   = help: set `edition = "2021"` in `Cargo.toml`
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
29	   = note: for more on editions, read https://doc.rust-lang.org/edition-guide
```

(https://github.com/rust-lang/rust/runs/4908271178?check_suite_focus=true)

### PR DESCRIPTION
**Pull request by ibraheemdev**
Resolves #93210

### CODE REVIEW

File: src/test/ui/did_you_mean/issue-93210-ignore-doc-hidden.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#[derive(Default)]
+pub struct B {
+    pub hello: i32,
+    pub bye: i32,
+}
+
+fn main() {
+    A::default().hey;
+    //~^ ERROR no field `hey` on type `A`
+    //~| NOTE unknown field
```
Comment (ibraheemdev) [inline src/test/ui/did_you_mean/issue-93210-ignore-doc-hidden.rs:17]: I'm not sure why this note is needed, it doesn't show up in the stderr output :thinking:

**Comment by nagisa** [comment]
I’m… not entirely sure this is a holistically a correct thing to do. Compiler diagnostics are not documentation. And the field _is_ accessible either way – it just takes another tool failing to account for `doc(hidden)` before its front-and-center before the user.

My intuition is that rustc itself should not look at the `doc` attributes to inform its own behaviour, much like it does not look at e.g. `#[rustfmt::*]` or `#[clippy::*]` attributes.

That said it seems like it already does account for `doc(hidden)` in at least some places such as e.g. https://github.com/rust-lang/rust/blob/ecf72996eda4f8af19b0ca7235c6f62e0245a313/compiler/rustc_mir_build/src/thir/pattern/usefulness.rs#L599 so I guess my intuition is invalid.

---

For the `Pin` case specifically, there's another potential problem – I suspect we suggest unstable fields when user is compiling with a stable compiler, or without a relevant feature enabled. That sounds like something that would want a fix given that the field cannot be accessed in that situation.

**Comment by m-ou-se** [comment]
> I’m… not entirely sure this is a holistically a correct thing to do. Compiler diagnostics are not documentation. And the field _is_ accessible either way – it just takes another tool failing to account for `doc(hidden)` before its front-and-center before the user.

`#[doc(hidden)]` is the only way and a very common way for libraries to mark something as 'not part of the pubilc api' when it has to be public for macro reasons. In the standard library we combine it with `#[unstable]`, but in libraries that need to work on stable Rust, there's not much more one can do that `#[doc(hidden)]` and adding a bunch of underscores in the name.

> I suspect we suggest unstable fields when user is compiling with a stable compiler, or without a relevant feature enabled. That sounds like something that would want a fix given that the field cannot be accessed in that situation.

It might be good to do the same thing as with enum variants: To also hide them when a field is part of an unstable feature that's not enabled.

**Comment by danielhenrymantilla** [comment]
I've added handling for unstability in my own branch: https://github.com/danielhenrymantilla/rust/commit/issue-93210

  - <sub><sup>(you can apply it here with `git pull https://github.com/danielhenrymantilla/rust issue-93210`)</sup></sub>

File: src/test/ui/did_you_mean/issue-93210-ignore-doc-hidden.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+#[derive(Default)]
+pub struct B {
+    pub hello: i32,
+    pub bye: i32,
+}
+
+fn main() {
+    A::default().hey;
```
Comment (danielhenrymantilla) [inline src/test/ui/did_you_mean/issue-93210-ignore-doc-hidden.rs:15]:
```suggestion
fn main() {
    A::default().hallo;
    A::default().hey;
```

`.hallo` may trigger similarly-named field suggestions, which may also be interesting behavior to test

**Comment by michaelwoerister** [comment]
Looks like there are design questions to be solved here, so r? rust-lang/diagnostics

**Review by davidtwco — APPROVED** [review]
I agree with @nagisa that it seems less than great that `#[doc(hidden)]` is used here.

It seems like we could either use a `pub(macro)` (can only be used in downstream crates when through one of this crate's macro), but I imagine that wouldn't really work with procedural macros; or a new attribute that the user can use to signal that something shouldn't be used in suggestions, rather than `#[doc(hidden)]`.

Regardless, it seems like `#[doc(hidden)]` is what we've got right now and we're using it for this in other places, so might as well do so here too. Implementation for this looks good - if you want to address @danielhenrymantilla's comment then feel free to, otherwise r=me (I'll do this in a day or two if there aren't any changes).

**Review by davidtwco — APPROVED** [review]

**Comment by danielhenrymantilla** [comment]
<sub><sup>Out-of-topic / meta comment: wow, since my commit was originally pushed (to my branch) quite a while ago, now that it has just been integrated in this PR, it nonetheless appears as occurring before I even made the very comment about it! 😆 (more seriously, I was confused by the lack of apparent changes in the "PR feed" near the bottom; I hope they somehow change that 😅).</sup></sub>

**Pull request merged**